### PR TITLE
optimize setLen

### LINKED ISSUE #6721: setLen(0) performance
given this snippet:
```Nim
import times
var a = new_seq[int]()

let tests = 500000000

var t = cpuTime()
for i in 0..tests:
    a.set_len(0)
echo "no check: " , cpuTime() - t

t = cpuTime()
for i in 0..tests:
    if len(a) != 0: a.set_len(0)
echo "check: ", cpuTime() - t
```

I see:
```
no check: 1.503536
check: 0.4542430000000002
```
I discovered this in an actual program where I am re-using a seq in a loop and setting the length to 0.
I haven't looked at the nim compiler code before this. I found `setLengthSeq` but I don't see why this is so problematic. Any pointers? Or am I doing something wrong?

### PR DESCRIPTION
**Pull request by brentp**
inline the call to setLengthSeq and avoid decref for types if ntfNoRefs

closes #6721 

This is my first attempted contribution to Nim so I especially appreciate any feedback. I think the check before the `forAllChildrenAux` is valid, but not positive.
 
with this code:
```Nim
import times

var N = 100000
var size = 200000
var a = newSeq[int](size)
var s:int

var t = cpuTime()
for i in 0..N:
  a.set_len(size)
  for k in 0..10:
    a.set_len(0)
    s += 1
  s += a.len

echo cpuTime() - t
echo s
```
I get:
```
$ nim c -d:release --gc:v2 -r seqbench.nim
3.158566
1100011

$ nim c -d:release --gc:refc -r seqbench.nim
3.16616
1100011

$ nim c -d:release --gc:boehm -r seqbench.nim
3.202457
1100011
```
before this change, I get:
```
$ nim c -d:release --gc:v2 -r seqbench.nim
16.951686
1100011

$ nim c -d:release --gc:refc -r seqbench.nim
17.303693
1100011

$ nim c -d:release --gc:boehm -r seqbench.nim
3.217406
1100011
```

### CODE REVIEW

**Comment by Varriount** [comment]
The inline pragma seems to be causing problems when the Nim runtime library feature is used. I don't believe it's really needed - the C/C++ compiler will likely inline that procedure if it thinks it is necessary.

File: lib/system/sysstr.nim [hunk 1]
```diff
@@ -259,7 +259,7 @@ proc incrSeqV2(seq: PGenericSeq, elemSize: int): PGenericSeq {.compilerProc.} =
     result.reserved = r
 
 proc setLengthSeq(seq: PGenericSeq, elemSize, newLen: int): PGenericSeq {.
-    compilerRtl.} =
+    compilerRtl, inline.} =
```
Comment (Araq) [inline lib/system/sysstr.nim]: Use ``inl`` here which is a variant of ``inline`` that knows about DLL boundaries.

**Comment by brentp** [comment]
Thanks for the feedback. I changed to inl. without inl or inline, the following code shows the slowdown:

```Nim
import times

var N = 1000000000

var a = new_seq[int]()
var s = 0

var t = cpuTime()
for i in 0..N:
  a.set_len(0)
  s += 1
echo s

echo cpuTime() - t

t = cpuTime()
s = 0
for i in 0..N:
  if a.len != 0: a.set_len(0)
  s += 1

echo s
echo cpuTime() - t
```

with the `inl` this gets the times much closer (1.95s without the check and 1.34 with the `len != 0` check.

**Comment by brentp** [comment]
bump. this is updated and squashed.

**Pull request merged**